src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/NodesDnApiAction.java

```diff
@@ -0,0 +1,145 @@
+import org.elasticsearch.rest.RestRequest.Method;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * This class implements CRUD operations to manage dynamic NodesDn. The primary usecase is targeted at cross-cluster where
+ * in node restart can be avoided by populating the coordinating cluster's nodes_dn values.
+ *
+ * The APIs are only accessible to SuperAdmin since the configuration controls the core application layer trust validation.
+ * By default the APIs are disabled and can be enabled by a YML setting - {@link ConfigConstants#OPENDISTRO_SECURITY_NODES_DN_DYNAMIC_CONFIG_ENABLED}
+ *
+ * The backing data is stored in {@link ConfigConstants#OPENDISTRO_SECURITY_CONFIG_INDEX_NAME} which is populated during bootstrap.
+ * For existing clusters, {@link com.amazon.opendistroforelasticsearch.security.tools.OpenDistroSecurityAdmin} tool can
+ * be used to populate the index.
+ *
+ * See {@link com.amazon.opendistroforelasticsearch.security.dlic.rest.api.NodesDnApiTest} for usage examples.
+ */
+public class NodesDnApiAction extends PatchableResourceApiAction {
+    public static final String STATIC_ES_YML_NODES_DN = "STATIC_ES_YML_NODES_DN";
+    private final List<String> staticNodesDnFromEsYml;
+
+    @Inject
+    public NodesDnApiAction(final Settings settings, final Path configPath, final RestController controller, final Client client,
+        final AdminDNs adminDNs, final ConfigurationRepository cl, final ClusterService cs,
+        final PrincipalExtractor principalExtractor, final PrivilegesEvaluator evaluator, ThreadPool threadPool, AuditLog auditLog) {
+        super(settings, configPath, controller, client, adminDNs, cl, cs, principalExtractor, evaluator, threadPool, auditLog);
+        this.staticNodesDnFromEsYml = settings.getAsList(ConfigConstants.OPENDISTRO_SECURITY_NODES_DN, Collections.emptyList());
+    }
+
+    @Override
+    protected void registerHandlers(RestController controller, Settings settings) {
+        if (settings.getAsBoolean(ConfigConstants.OPENDISTRO_SECURITY_NODES_DN_DYNAMIC_CONFIG_ENABLED, false)) {
+            controller.registerHandler(Method.GET, "/_opendistro/_security/api/nodesdn/{name}", this);
+            controller.registerHandler(Method.GET, "/_opendistro/_security/api/nodesdn/", this);
+            controller.registerHandler(Method.DELETE, "/_opendistro/_security/api/nodesdn/{name}", this);
+            controller.registerHandler(Method.PUT, "/_opendistro/_security/api/nodesdn/{name}", this);
+            controller.registerHandler(Method.PATCH, "/_opendistro/_security/api/nodesdn/", this);
+            controller.registerHandler(Method.PATCH, "/_opendistro/_security/api/nodesdn/{name}", this);
+        }
+    }
+
+    @Override
+    protected void handleApiRequest(RestChannel channel, RestRequest request, Client client) throws IOException {
+        if (!isSuperAdmin()) {
```

Comment:
Can the API be blocked from RestApiPrivilegeEvaluator instead of doing this here? 
https://github.com/opendistro-for-elasticsearch/security/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/RestApiPrivilegesEvaluator.java#L213

Reply:
I tried that approach - however that handles RBAC which doesn't support special casing for admin DN (AdminDN is required since this is sensitive setting parallel to cert hot-reload)

Reply:
AdminDN is already injected in the constructor and user can be determined from the thread pool context. How differently are they handled in the api action ?